sciope/inference/rep_smc_abc.py

```diff
@@ -0,0 +1,187 @@
+    * use_logger    			(whether logging is enabled or disabled)
+
+    Methods:
+    * infer 					(perform parameter inference)
+    """
+
+    def __init__(self, data, sim, prior_function, perturbation_kernel,
+                 summaries_function=bs.Burstiness(),
+                 distance_function=euc.EuclideanDistance(),
+                 summaries_divisor=None, use_logger=False):
+
+        self.name = 'SMC-ABC'
+        super(ReplenishmentSMCABC, self).__init__(self.name, data, sim, use_logger)
+
+        self.prior_function = prior_function
+        self.summaries_function = summaries_function
+        self.distance_function = distance_function
+        self.summaries_divisor = summaries_divisor
+        self.perturbation_kernel = perturbation_kernel
+
+        if self.use_logger:
+            self.logger = ml.SciopeLogger().get_logger()
+            self.logger.info("Sequential Monte-Carlo Approximate Bayesian Computation initialized")
+
+    def _simulate_N(self, prior, N):
+
+        params_delayed = prior.draw(N)
+        params, = dask.compute(params_delayed, num_workers = 10)
+        params = np.vstack(params).reshape(N, -1)
+
+        results = []
+        for i in range(params.shape[0]):
+            current_param = params[i,:]
+            lazy_results = dask.delayed(self._simulate_distance)(current_param)
+            results.append(lazy_results)
+
+        computed_results, = dask.compute(results)
+
+        return np.asarray(computed_results), params
+
+    def _simulate_distance(self, param):
+        res = self.summaries_function(self.sim(param))
+        res_obs = self.summaries_function(self.data)
+        return self.distance_function.compute(res_obs, res)
+
+    def infer(self, num_samples, alpha, R_trial = 10, c = 0.01, p_min = 0.05):
+        """Performs SMC-ABC.
+
+        Parameters
+        ----------
+        num_samples : int
+            The number of required accepted samples
+        alpha : float
+            Culling percentage
+        R_trial : int
+            Number of perturbs per replenishment
+
+        Returns
+        -------
+        dict
+            Keys
+            'accepted_samples: The accepted parameter values',
+            'distances: Accepted distance values'
+        """
+        def perturb_resample(p, cur_distance, R, tol):
+            p_acc = 0
+            n_moves = 0
+            new_distance = cur_distance
+            for _ in range(R):
+                z = self.perturbation_kernel.rvs(p)
+                num = prior.pdf(z)
+                # Metropolis Perturbation
+                if num > 0:
+                    num = num * self.perturbation_kernel.pdf(p.reshape(1,-1), z.reshape(1,-1))[0,0,]
+                    dnm = prior.pdf(p) * self.perturbation_kernel.pdf(z.reshape(1,-1), p.reshape(1,-1))[0,0]
+                    weight = num/dnm
+                    if np.random.rand() < min(1, weight):
+                        distance = self._simulate_distance(z)
```

Comment:
need to be relplaced if `sciope.core.get_graph_[<chunked>,<unchunked>]` are used instead